### PR TITLE
Mathtokens

### PR DESCRIPTION
**Pull request by mhrmm**
@sbhaktha Sending this to you because you're currently the most familiar with the random forest stuff. Two things: (1) an optimization of the RandomForest.outcomeDistribution method, (2) some additional math-specific features for parsing sentences with inline math.

### CODE REVIEW

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/RandomForest.scala [hunk 1]
```diff
@@ -35,23 +35,32 @@ case class RandomForest(allOutcomes: Seq[Int], decisionTrees: Seq[DecisionTree])
...
-      decisionTree.outcomeDistribution(featureVector)._1
+    // determines whether we've seen enough of one outcome to conclude that
+    // it's the majority outcome
+    def isConfidentEnoughAboutHistogram(histogram: Map[Int, Int]): Boolean = {
+      histogram.values.max > decisionTrees.size / 2
+    }
+    var decisionTreeIndex = 0
+    var continueLoop: Boolean = true
+    var unnormalizedOutcomeDistribution = Map[Int, Int]()
+    while (decisionTreeIndex < decisionTrees.size && continueLoop) {
```
Comment (sbhaktha) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/RandomForest.scala]: This makes sense, but doesn't read very Scala-y to me.
Suggestion- something like:
```
  var continueLoop = true
  var unnormalizedOutcomeDistribution = Map[Int, Int]()
  for {
    decisionTree <- decisionTrees
    if (continueLoop)
  }  {
    val outcome = decisionTree.classify(featureVector)
    unnormalizedOutcomeDistribution =
        unnormalizedOutcomeDistribution.updated(
          outcome._1,
          1 + unnormalizedOutcomeDistribution.getOrElse(outcome._1, 0)
        )
    continueLoop = !isConfidentEnoughAboutHistogram(unnormalizedOutcomeDistribution)
  }
```

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/RandomForest.scala [hunk 2]
```diff
@@ -35,23 +35,32 @@ case class RandomForest(allOutcomes: Seq[Int], decisionTrees: Seq[DecisionTree])
...
+      val nextDecisionTree = decisionTrees(decisionTreeIndex)
+      val outcome: (Int, Option[Justification]) = nextDecisionTree.classify(featureVector)
+      unnormalizedOutcomeDistribution =
+        unnormalizedOutcomeDistribution.updated(
+          outcome._1,
+          1 + unnormalizedOutcomeDistribution.getOrElse(outcome._1, 0)
+        )
+      continueLoop = !isConfidentEnoughAboutHistogram(unnormalizedOutcomeDistribution)
```
Comment (sbhaktha) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/RandomForest.scala]: Perhaps call this only if you've already traversed at least half the decision trees? Right now you have the decision tree index. If you do refactor with a `for` comprehension, you could perhaps use a `zipWithIndex`.

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/core/TokenTagger.scala [hunk 3]
```diff
@@ -64,7 +64,18 @@ case object LexicalPropertiesTagger extends TokenTagger {
     } map {
       _ => 'existsNum
     }
-    (Seq(firstLetterCapital, existsCapital, allCaps, existsNumber) map { maybeFeat =>
+    def parseDouble(s: String) = try { Some(s.toDouble) } catch { case _: Throwable => None }
+    val looksLikeDouble = parseDouble(tokStr) map {
+      _ => 'looksLikeDouble
+    }
+    val looksLikeGeoLabel =
```
Comment (sbhaktha) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/core/TokenTagger.scala]: `Geo` is almost always abbreviation for "geographic". I really wondered for an instant what we're doing with geo stuff. Perhaps rename to `Geom`?

**Comment by sbhaktha** [comment]
A couple of comments. Looks good overall.

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/RandomForest.scala (hunk 2 above)
Comment (mhrmm) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/RandomForest.scala]: I'd tried, but profiling indicated this doesn't provide any speedup (and this way is a bit simpler).

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/core/TokenTagger.scala (hunk 3 above)
Comment (mhrmm) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/core/TokenTagger.scala]: Yeah, somebody else mentioned that too. :) I'll change to geom.

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/RandomForest.scala (hunk 1 above)
Comment (mhrmm) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/RandomForest.scala]: I like that change. Thanks!

**Comment by mhrmm** [comment]
Revised!

**Pull request merged**